lens-cube/src/main/java/org/apache/lens/cube/parse/CandidateCoveringSetsResolver.java

```diff
@@ -82,73 +82,80 @@ private void updateFinalCandidates(List<List<UnionCandidate>> jcs) {
     }
   }
 
-  private void resolveRangeCoveringFactSet(CubeQueryContext cubeql, TimeRange range,
+  private boolean isCandidateCoveringTimeRanges(UnionCandidate uc, List<TimeRange> ranges) {
+    for (Iterator<TimeRange> itr = ranges.iterator(); itr.hasNext(); ) {
+      TimeRange range = itr.next();
+      if (!CandidateUtil.isTimeRangeCovered(uc.getChildCandidates(), range.getFromDate(), range.getToDate())) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private void pruneUnionCandidatesNotCoveringAllRanges(List<UnionCandidate> ucs, List<TimeRange> ranges) {
+    for (Iterator<UnionCandidate> itr = ucs.iterator(); itr.hasNext(); ) {
+      UnionCandidate uc = itr.next();
+      if (!isCandidateCoveringTimeRanges(uc, ranges)) {
+        itr.remove();
+      }
+    }
+  }
+
+  private void resolveRangeCoveringFactSet(CubeQueryContext cubeql, List<TimeRange> ranges,
                                            Set<QueriedPhraseContext> queriedMsrs) throws LensException {
-    // All facts
-    List<Candidate> allCandidates = cubeql.getCandidateSet();
-    // Partially valid facts
+    // All Candidates
+    List<Candidate> allCandidates = new ArrayList<Candidate>(cubeql.getCandidates());
+    // Partially valid candidates
     List<Candidate> allCandidatesPartiallyValid = new ArrayList<>();
     for (Candidate cand : allCandidates) {
       // Assuming initial list of candidates populated are StorageCandidate
       if (cand instanceof StorageCandidate) {
         StorageCandidate sc = (StorageCandidate) cand;
-        if (CandidateUtil.isValidForTimeRange(sc, range)) {
-          List<Candidate> one = new ArrayList<Candidate>(Arrays.asList(sc));
-          unionCandidates.add(new UnionCandidate(one));
-        } else if (CandidateUtil.isPartiallyValidForTimeRange(sc, range)) {
-          allCandidatesPartiallyValid.add(sc);
+        if (CandidateUtil.isValidForTimeRanges(sc, ranges)) {
+          List<Candidate> one = new ArrayList<Candidate>(Arrays.asList(CandidateUtil.cloneStorageCandidate(sc)));
+          unionCandidates.add(new UnionCandidate(one, "uc" + unionCandidatealiasCounter++));
```

Comment:
Can UnionCandidate have just one child ? Should this be StorageCandidate Instead ?